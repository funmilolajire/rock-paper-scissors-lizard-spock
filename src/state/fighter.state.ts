import { createState, useState, State, DevTools } from '@hookstate/core';
import { outcome } from './utils/rules';
import { DevToolsInitialize } from '@hookstate/devtools';

export const fighters = ["rock", "paper", "scissors", "lizard", "spock"];

const playerFighter = createState('');
const houseFighter = createState('');

const wrapState = (state: State<string>) => ({
    get: () => state.value,
    select: (name: string) => state.set(prev => prev = name),
})

//to use outside of a component
export const accessPlayerFighter = () => wrapState(playerFighter);
export const accessHouseFighter = () => wrapState(houseFighter);

//to use within a component
export const usePlayerFighter = () => wrapState(useState(playerFighter));
export const useHouseFighter = () => wrapState(useState(houseFighter));

//devtools
DevTools(playerFighter).label('player-fighter')
DevTools(houseFighter).label('house-fighter')
DevToolsInitialize({
    monitored: ['player-fighter', 'house-fighter'],
    callstacksDepth: 30
})
