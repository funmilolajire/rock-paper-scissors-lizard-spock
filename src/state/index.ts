export * from './modal.state';
export * from './score.state';
export * from './fighter.state';