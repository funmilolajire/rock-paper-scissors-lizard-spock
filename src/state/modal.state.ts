import { createState, useState, State, DevTools } from '@hookstate/core';
import { DevToolsInitialize } from '@hookstate/devtools';

const modalState = createState(false);
const wrapState = (state: State<boolean>) => ({
    get: () => state.value,
    open: () => state.set(true),
    close: () => state.set(false)
})

//to use outside of a component
export const accessModalState = () => wrapState(modalState);

//to use within a component
export const useModalState = () => wrapState(useState(modalState));

//devtools
DevTools(modalState).label('modal-state')
DevToolsInitialize({
    monitored: ['modal-state'],
    callstacksDepth: 30
})
