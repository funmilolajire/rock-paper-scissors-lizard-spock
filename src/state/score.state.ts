import { createState, useState, State, DevTools } from '@hookstate/core';
import { DevToolsInitialize } from '@hookstate/devtools';
import { Persistence } from '@hookstate/persistence'


const playerState = createState(0);
const houseState = createState(0);

const wrapState = (state: State<number>) => ({
    get: () => state.value,
    increment: () => state.set(prev => prev + 1),
    clear: () => state.set(0)
})

//to use from outside a component
export const accessPlayerState = () => wrapState(playerState.attach(Persistence('playerScore')))
export const accessHouseState = () => wrapState(houseState.attach(Persistence('houseScore')))

//to use from within a component
export const usePlayerState = () => wrapState(useState(playerState).attach(Persistence('playerScore')))
export const useHouseState = () => wrapState(useState(houseState).attach(Persistence('houseScore')))

//devtools
DevTools(playerState).label('player-state')
DevTools(houseState).label('house-state')
DevToolsInitialize({
    monitored: ['player-state', 'house-state'],
    callstacksDepth: 30
})
