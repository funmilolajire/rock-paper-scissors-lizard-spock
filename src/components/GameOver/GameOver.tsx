import { useHouseFighter, usePlayerFighter } from "../../state"

export const GameOver = ({ gameOver }: { gameOver: string }) => {
    const playerFighter = usePlayerFighter()
    const houseFighter = useHouseFighter()
    const playAgain = (): void => {
        playerFighter.select('')
        houseFighter.select('')
    }
    return (
        <div className="GameOver">
            <h3>You {gameOver}</h3>
            <button onClick={playAgain}>Play Again</button>
        </div>
    )
}