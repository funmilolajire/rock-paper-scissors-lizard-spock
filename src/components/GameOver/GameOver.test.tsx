import { mount } from 'enzyme';
import { GameOver } from './GameOver';
import { accessPlayerFighter, accessHouseFighter } from '../../state/fighter.state';

describe('gameover component', () => {
    let props = {
        gameOver: "win"
    }
    let component = mount(<GameOver {...props} />)
    test('make a snapshot of gameover component', () => {
        expect(component).toMatchSnapshot()
    })
    test('it restarts game on click', () => {
        let restart = component.find('.GameOver button');
        restart.simulate('click')
        expect(accessPlayerFighter().get()).toBeFalsy()
        expect(accessHouseFighter().get()).toBeFalsy()
    })
})
