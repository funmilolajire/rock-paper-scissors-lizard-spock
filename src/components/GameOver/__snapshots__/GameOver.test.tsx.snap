// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gameover component make a snapshot of gameover component 1`] = `
<GameOver
  gameOver="win"
>
  <div
    className="GameOver"
  >
    <h3>
      You 
      win
    </h3>
    <button
      onClick={[Function]}
    >
      Play Again
    </button>
  </div>
</GameOver>
`;
